signature/src/hazmat.rs

```diff
@@ -0,0 +1,49 @@
+//! Hazardous Materials: low-level APIs which can be insecure if misused.
+//!
+//! The traits in this module are not generally recommended, and should only be
+//! used in special cases where they are specifically needed.
+//!
+//! Using them incorrectly can introduce security vulnerabilities. Please
+//! carefully read the documentation before attempting to use them.
+//!
+//! To use them, enable the `hazmat-preview` crate feature. Note that this
+//! feature is semi-unstable and not subject to regular 1.x SemVer guarantees.
+//! However, any breaking changes will be accompanied with a minor version bump.
+
+use crate::{Error, Signature};
+
+/// Sign the provided message prehash, returning a digital signature.
+pub trait PrehashSigner<S: Signature> {
+    /// Attempt to sign the given message digest, returning a digital signature
+    /// on success, or an error if something went wrong.
+    ///
+    /// The `prehash` parameter should be the output of a secure cryptographic
+    /// hash function.
+    ///
+    /// This API takes a `prehash` byte slice as there can potentially be many
+    /// compatible lengths for the message digest for a given concrete signature
+    /// algorithm.
+    ///
+    /// Allowed lengths are algorithm-dependent and up to a particular
+    /// implementation to decide.
+    fn try_sign_prehash(&self, prehash: &[u8]) -> Result<S, Error>;
```

Comment:
I thought about a few different alternatives to `prehash: &[u8]` (i.e. a byte slice), including:

- Adding a const generic parameter to the trait, which would permit overlapping impls for different sizes
- Using `generic-array`

While both of those are worth considering, algorithms like DSA and RSASSA-PKCS#1v15/PSS both currently use `num-bigint-dig` with heap-allocated bignums that don't know their own size at a type level.

So I think a slice really is the only option that's general enough.

Reply:
We could introduce two separate methods: array and slice based. The latter would have a default impl in terms of the former. The mentioned algorithm implementations would overwrite the slice-based method and implement the array-based one using slice-based.

Reply:
The only algorithm that would benefit for now is ECDSA, and that would need `generic-array` so the implementation in the `ecdsa` crate can be generic around field sizes.

However, using the slice-based API means we can implement hash truncation / zero padding in one place, which would make it easier to support some exotic combinations of curve and hash functions, such as ECDSA/P-256 with SHA384, or ECDSA/P-384 with SHA-256.